libtransmission/torrent-magnet.cc

```diff
@@ -263,27 +263,16 @@ void on_have_all_metainfo(tr_torrent* tor)
     auto error = tr_error{};
     auto& m = tor->incomplete_metadata;
     TR_ASSERT(m);
-    if (use_new_metainfo(tor, &error))
-    {
-        m.reset();
-    }
-    else /* drat. */
-    {
-        auto const n = m->piece_count;
-
-        m->pieces_needed = create_all_needed(n);
 
+    if (!use_new_metainfo(tor, &error)) /* drat. */
+    {
         auto msg = std::string_view{ error && !std::empty(error.message()) ? error.message() : "unknown error" };
         tr_logAddWarnTor(
             tor,
-            fmt::format(
-                tr_ngettext(
-                    "Couldn't parse magnet metainfo: '{error}'. Redownloading {piece_count} piece",
-                    "Couldn't parse magnet metainfo: '{error}'. Redownloading {piece_count} pieces",
-                    n),
-                fmt::arg("error", msg),
-                fmt::arg("piece_count", n)));
+            fmt::format("Couldn't parse magnet metainfo: '{error}'. Redownloading metadata", fmt::arg("error", msg)));
     }
+
+    m.reset();
```

Comment:
I'm thinking that this missing reset() was possibly why the rollback of 5791 was working in #6272: that rollback was using a different `tr_incomplete_metadata` each time, instead of a unique instance, so it was able to recover on errors.

So possibly the original regression of the metadata downloads wasn't on July 14th with ca4cb1a675fb951f6ca975e8bb3d4de1ae25ad57, but was on July 15th with c805e7533f115850e41b116367e8a713c0a97949.

Anyway, you're trying to address both, so it's fine.